lib/stdlib/src/shell_docs.erl

```diff
@@ -575,7 +582,65 @@ get_type_doc(Module, Type, Arity) ->
       Module :: module(),
       Docs :: docs_v1().
 render_type(Module, D) ->
-    render_type(Module, D, #{}).
+  render_type(Module, D, #{}).
+
+%% extract AST raw type specifications.
+extract_type_specs(Module) ->
+  maybe
+    Path ?= find_path(Module),
+    true ?= non_existing =/= Path,
+    {ok, {_ModName,
+            [{debug_info,
+                {debug_info_v1,erl_abstract_code,
+                   {AST, _Opts}}}]}} ?= beam_lib:chunks(Path, [debug_info]),
+
+    %% the mapping keys 'type', 'function', and 'callback' correspond
+    %% to existing EEP-48 {**Kind**, Name, Arity} format, where Kind
+    %% ranges over these values. This is needed to differentiate
+    %% function, callback, and types when looking up their type specs
+    Acc = #{type => #{}, 'function' => #{}, 'callback' => #{}},
+    lists:foldl(fun filter_exported_types/2, Acc, AST)
+  else
+    {error, _} -> #{};
+    non_existing -> #{};
+    {error,beam_lib,{file_error,_,enoent}} -> #{}
+  end.
+
+find_path(Module) ->
+    case code:which(Module) of
+        preloaded ->
+            PreloadedPath = filename:join(code:lib_dir(erts),"ebin"),
+            PreloadedPath ++ "/" ++ atom_to_list(Module);
```

Comment:
```suggestion
            filename:join(PreloadedPath, atom_to_list(Module) ++ ".beam");
```
Or windows will/may break. Although not required by the current `beam_lib:chunks/2` API, I think it is prudent to add the ".beam" extension.